File: backend/bin/migrate_mdx.rs
use std::path::{Path, PathBuf};
use std::fs;
use std::env;
use sqlx::PgPool;
use chrono::DateTime;
use uuid::Uuid;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Get database URL from environment
    let database_url = env::var("DATABASE_URL")
        .unwrap_or_else(|_| "postgresql://blog_user:blog_password@localhost:5432/blog_db".to_string());

    let blog_dir = env::var("FRONTEND_BLOG_DIR")
        .unwrap_or_else(|_| "./frontend/data/blog".to_string());

    println!("================================");
    println!("MDX 文件迁移工具");
    println!("================================");
    println!("数据库: {}", database_url);
    println!("文章目录: {}", blog_dir);
    println!("");

    // Connect to database
    println!("1. 连接数据库...");
    let pool = PgPool::connect(&database_url).await?;
    println!("✓ 数据库连接成功");
    println!("");

    // Count MDX files
    println!("2. 扫描 MDX 文件...");
    let mdx_files = find_mdx_files(&blog_dir)?;
    println!("✓ 找到 {} 个 MDX 文件", mdx_files.len());
    println!("", );

    // Migrate files
    println!("3. 开始迁移...");
    let mut success_count = 0;
    let mut error_count = 0;

    for (i, mdx_path) in mdx_files.iter().enumerate() {
        print!("  [{}/{}] {} ... ", i + 1, mdx_files.len(), mdx_path.display());

        match migrate_file(&pool, &mdx_path).await {
            Ok(_) => {
                println!("✓");
                success_count += 1;
            }
            Err(e) => {
                println!("✗ {}", e);
                error_count += 1;
            }
        }
    }

    println!("");
    println!("================================");
    println!("迁移完成！");
    println!("================================");
    println!("成功: {} 个", success_count);
    println!("失败: {} 个", error_count);
    println!("");

    // Verify results
    println!("4. 验证结果...");
    let categories_count: i64 = sqlx::query_scalar("SELECT COUNT(*) FROM categories")
        .fetch_one(&pool).await?;
    let tags_count: i64 = sqlx::query_scalar("SELECT COUNT(*) FROM tags")
        .fetch_one(&pool).await?;
    let posts_count: i64 = sqlx::query_scalar("SELECT COUNT(*) FROM posts")
        .fetch_one(&pool).await?;

    println!("  - 分类: {}", categories_count);
    println!("  - 标签: {}", tags_count);
    println!("  - 文章: {}", posts_count);
    println!("");

    Ok(())
}

fn find_mdx_files(dir: &str) -> Result<Vec<PathBuf>, Box<dyn std::error::Error>> {
    let mut files = Vec::new();
    let blog_path = Path::new(dir);

    if blog_path.exists() {
        fn visit_dirs(dir: &Path, files: &mut Vec<PathBuf>) {
            if dir.is_dir() {
                for entry in fs::read_dir(dir).unwrap() {
                    let entry = entry.unwrap();
                    let path = entry.path();
                    if path.is_dir() {
                        visit_dirs(&path, files);
                    } else if path.extension().and_then(|s| s.to_str()) == Some("mdx") {
                        files.push(path);
                    }
                }
            }
        }
        visit_dirs(blog_path, &mut files);
        files.sort();
    }

    Ok(files)
}

async fn migrate_file(pool: &PgPool, mdx_path: &Path) -> Result<(), String> {
    // Read file content
    let content = fs::read_to_string(mdx_path)
        .map_err(|e| format!("读取文件失败: {}", e))?;

    // Parse frontmatter and content
    let (title, slug, date, category, summary, body) = parse_mdx(&content, mdx_path)?;

    // Get category ID
    let category_id: Option<Uuid> = sqlx::query_scalar(
        "SELECT id FROM categories WHERE slug = $1 OR name = $1 LIMIT 1"
    )
    .bind(&category)
    .fetch_optional(pool)
    .await
    .map_err(|e| format!("查询分类失败: {}", e))?;

    // Check if post already exists
    let exists: bool = sqlx::query_scalar(
        "SELECT EXISTS(SELECT 1 FROM posts WHERE slug = $1)"
    )
    .bind(&slug)
    .fetch_one(pool)
    .await
    .map_err(|e| format!("检查文章是否存在失败: {}", e))?;

    if exists {
        return Err("文章已存在".to_string());
    }

    // Insert post
    sqlx::query(
        r#"
        INSERT INTO posts (
            slug, title, summary, content, status,
            published_at, category_id, view_count, like_count, comment_count
        ) VALUES ($1, $2, $3, $4, 'published', $5, $6, 0, 0, 0)
        "#
    )
    .bind(&slug)
    .bind(&title)
    .bind(&summary)
    .bind(&body)
    .bind(&date)
    .bind(&category_id)
    .execute(pool)
    .await
    .map_err(|e| format!("插入文章失败: {}", e))?;

    Ok(())
}

fn parse_mdx(content: &str, path: &Path) -> Result<(String, String, DateTime<chrono::Utc>, String, String, String), String> {
    let mut lines = content.lines().peekable();

    // Check for frontmatter delimiter
    if lines.peek() != Some(&"---") {
        return Err("缺少 frontmatter".to_string());
    }
    lines.next();

    // Parse frontmatter
    let mut title = String::new();
    let mut date_str = String::new();
    let mut category = String::new();
    let mut summary = String::new();
    let mut tags = Vec::new();

    loop {
        match lines.next() {
            Some("---") => break,
            Some(line) if line.starts_with("title:") => {
                title = line.replacen("title:", "", 1)
                    .trim().trim_matches('"').trim_matches('\'').to_string();
            }
            Some(line) if line.starts_with("date:") => {
                date_str = line.replacen("date:", "", 1).trim().to_string();
            }
            Some(line) if line.starts_with("category:") => {
                category = line.replacen("category:", "", 1)
                    .trim().trim_matches('"').trim_matches('\'').to_string();
            }
            Some(line) if line.starts_with("summary:") => {
                summary = line.replacen("summary:", "", 1)
                    .trim().trim_matches('"').trim_matches('\'').to_string();
            }
            Some(line) if line.starts_with("tags:") => {
                // Parse tags array - simplified
                let tags_str = line.replacen("tags:", "", 1);
                // TODO: Parse tags properly
            }
            Some(_) => {}
            None => return Err("Frontmatter 未正确关闭".to_string()),
        }
    }

    // Generate defaults
    if title.is_empty() {
        title = path.file_stem()
            .and_then(|s| s.to_str())
            .unwrap_or("untitled")
            .to_string();
    }

    let slug = title.to_lowercase()
        .replace(' ', "-")
        .replace('/', "-")
        .chars()
        .filter(|c| c.is_alphanumeric() || *c == '-' || *c == '_')
        .collect::<String>();

    let date = if date_str.is_empty() {
        chrono::Utc::now()
    } else {
        DateTime::parse_from_rfc3339(&format!("{}T00:00:00Z", date_str))
            .or_else(|_| DateTime::parse_from_rfc3339(&format!("{}T00:00:00+00:00", date_str)))
            .map(|dt| dt.with_timezone(&chrono::Utc))
            .unwrap_or_else(|_| chrono::Utc::now())
    };

    if category.is_empty() {
        category = "computer-science".to_string();
    }

    // Collect body content
    let body = lines.collect::<Vec<&str>>().join("\n");

    Ok((title, slug, date, category, summary, body))
}

File: scripts/archive/migrate_mdx.py
#!/usr/bin/env python3
"""MDX 文件迁移工具 - 通过 API 导入"""

import os
import re
import json
import requests
from pathlib import Path
from datetime import datetime
from zoneinfo import ZoneInfo

# 配置
API_URL = "http://127.0.0.1:3000/v1"
BLOG_DIR = "./frontend/data/blog"
DEFAULT_CATEGORY = "computer-science"

def find_mdx_files(directory):
    """递归查找所有 MDX 文件"""
    mdx_files = []
    for root, dirs, files in os.walk(directory):
        for file in files:
            if file.endswith('.mdx'):
                mdx_files.append(os.path.join(root, file))
    return sorted(mdx_files)

def parse_mdx(filepath):
    """解析 MDX 文件"""
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()

    # 分离 frontmatter 和正文
    lines = content.split('\n')

    if not lines or lines[0] != '---':
        return None

    frontmatter_lines = []
    body_lines = []
    in_frontmatter = True
    dashes_count = 0

    for line in lines[1:]:
        if line == '---' and dashes_count == 0:
            dashes_count += 1
            in_frontmatter = False
            continue
        if in_frontmatter:
            frontmatter_lines.append(line)
        else:
            body_lines.append(line)

    # 解析 frontmatter
    frontmatter = {}
    current_key = None
    current_value = []

    for line in frontmatter_lines:
        match = re.match(r'^(\w+):\s*(.*)$', line)
        if match:
            if current_key:
                frontmatter[current_key] = '\n'.join(current_value).strip()
            current_key = match.group(1)
            current_value = [match.group(2)]
        elif current_key:
            current_value.append(line)

    if current_key:
        frontmatter[current_key] = '\n'.join(current_value).strip()

    # 提取信息
    title = frontmatter.get('title', Path(filepath).stem)
    slug = title.lower().replace(' ', '-').replace('/', '-')
    date_str = frontmatter.get('date', datetime.now().strftime('%Y-%m-%d'))
    category = frontmatter.get('category', DEFAULT_CATEGORY).strip('"\'')
    summary = frontmatter.get('summary', '')

    # 解析日期
    try:
        published_at = datetime.strptime(date_str, '%Y-%m-%d').isoformat() + 'Z'
    except:
        published_at = datetime.now().isoformat() + 'Z'

    # 标签（如果有）
    tags_str = frontmatter.get('tags', '[]')
    try:
        tags = json.loads(tags_str) if isinstance(tags_str, str) else []
    except:
        tags = []

    body = '\n'.join(body_lines).strip()

    return {
        'title': title,
        'slug': slug,
        'summary': summary,
        'content': body,
        'published_at': published_at,
        'category': category,
        'tags': tags
    }

def create_category(name, api_key):
    """创建分类（如果不存在）"""
    slug = name.lower().replace(' ', '-')

    # 检查是否已存在
    response = requests.get(f"{API_URL}/categories/{slug}")
    if response.status_code == 200:
        return response.json()['id']

    # 创建新分类
    data = {
        'slug': slug,
        'name': name,
        'description': f'{name} articles'
    }

    response = requests.post(
        f"{API_URL}/admin/categories",
        json=data,
        headers={'Authorization': f'Bearer {api_key}'}
    )

    if response.status_code in [200, 201]:
        return response.json()['id']
    return None

def migrate_file(filepath, api_key):
    """迁移单个文件"""
    parsed = parse_mdx(filepath)
    if not parsed:
        print(f"  ✗ 无法解析: {filepath}")
        return False

    # 确保分类存在
    category_id = create_category(parsed['category'], api_key)
    if not category_id:
        print(f"  ⚠ 无法创建分类: {parsed['category']}")

    # 创建文章
    data = {
        'slug': parsed['slug'],
        'title': parsed['title'],
        'summary': parsed['summary'],
        'content': parsed['content'],
        'published_at': parsed['published_at'],
        'category_id': category_id,
        'status': 'published'
    }

    response = requests.post(
        f"{API_URL}/admin/posts",
        json=data,
        headers={'Authorization': f'Bearer {api_key}'}
    )

    if response.status_code in [200, 201]:
        return True
    else:
        print(f"  ✗ 失败: {response.text}")
        return False

def main():
    print("=" * 40)
    print("MDX 文件迁移工具")
    print("=" * 40)
    print("")

    # 检查 API
    print("1. 检查 API 连接...")
    try:
        response = requests.get(f"{API_URL}/healthz")
        if response.status_code != 200:
            print("  ✗ API 不可用")
            return
        print("  ✓ API 连接正常")
    except Exception as e:
        print(f"  ✗ 无法连接到 API: {e}")
        return

    # 查找 MDX 文件
    print("2. 扫描 MDX 文件...")
    mdx_files = find_mdx_files(BLOG_DIR)
    print(f"  ✓ 找到 {len(mdx_files)} 个 MDX 文件")
    print("")

    # 注意：这需要管理员权限
    print("⚠ 注意: 此脚本需要管理员 API 密钥")
    print("如果还没有创建管理员用户，请先创建")
    print("")

    # TODO: 这里需要先实现认证，或者跳过认证直接插入数据库
    print("由于需要认证，此脚本暂时无法执行")
    print("建议使用数据库直接插入的方式")
    print("")

    print("3. 将执行迁移...")
    print("   (暂时跳过)")

if __name__ == '__main__':
    main()
